service/src/event/order_events.rs

```diff
@@ -47,25 +47,31 @@ impl OrderEvents {
                 (e.cid(), e.deliverable())
             }));
         let mut deliverable = Vec::with_capacity(candidate_events.len());
-        candidate_events.retain(|(e, h)| {
+        let mut remaining_candidates = Vec::with_capacity(candidate_events.len());
+
+        while !candidate_events.is_empty() {
+            let (e, h) = candidate_events.pop().unwrap();
```

Comment:
the unwrap() is safe because of the is_empty check in the while loop condition

Reply:
How about `while let Some((e,h)) = candidate_events.pop() {`